# Deciding where open a Baseball shop in Toronto

Marcus Lopes

March 30, 2019

## Table of Contents

1. Introduction
2. Data
3. Methodology
4. Results
5. Discussion
6. Conclusion

## 1. Introduction

### 1.1. Background
The neighborhoods in a city are very diversificated, but we can cluster them in a way that are possible to identify patterns in what venues are most common in each group of similar neighborhoods. 

### 1.2. Problem
This will be especially important to us, because we are contracted to determine where should be open a baseball’s stuff shop in the city of Toronto. So we will identify where are the neighborhoods that will have a stablished potencial consumers based in the atual venues nearest. 

### 1.3. Interest
This will be important to give the entrepreneur the best chances to have succed with him shop and have a better return for him investment.

## 2. Data

The data we will use to solve the problem presented above will be the table contained in the Wikipedia page https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M in wich we will extract the postal codes, boroughs and neighborhoods of Toronto, Canada. And with geospatial data provided in the csv http://cocl.us/Geospatial_data from the Capstone Project course of Coursera and IBM we will append the latitude and longitude from postal codes to our table previously extracted. Finally we will use the location data from Foursquare to identify wich categories of venues are more contained in each neighborhood to then clustering this data and so identify the principals caregories of venues contained in each neighborhood’s cluster. This will permit make a decision about where open the baseball's stuff store.

## 3. Methodology

Segmenting and Clustering Neighborhoods in Toronto.

### 3.1. Catching neighborhoods data from source

In [1]:
# import libraries

import requests
from bs4 import BeautifulSoup
import pandas as pd

In [2]:
# wikipedia page that contains the postal codes, boroughs and neighborhoods from Toronto

wikipedia_link = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wikipedia_page = requests.get(wikipedia_link)
html_doc = wikipedia_page.text
soup = BeautifulSoup(html_doc, 'html.parser')

In [3]:
# html from wikipedia page

soup

<!DOCTYPE html>

<html class="client-nojs" dir="ltr" lang="en">
<head>
<meta charset="utf-8"/>
<title>List of postal codes of Canada: M - Wikipedia</title>
<script>document.documentElement.className = document.documentElement.className.replace( /(^|\s)client-nojs(\s|$)/, "$1client-js$2" );</script>
<script>(window.RLQ=window.RLQ||[]).push(function(){mw.config.set({"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":false,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":890001695,"wgRevisionId":890001695,"wgArticleId":539066,"wgIsArticle":true,"wgIsRedirect":false,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Communications in Ontario","Postal codes in Canada","Toronto","Ontario-related lists"],"wgBreakFrames":false,"wgPageContentLanguage":"en","wgPageContentModel":"wikitext","wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","

### Now let's extract table from html...

We will use just the table containing the postal codes, boroughs and neighborhoods data from wikipedia page.

In [4]:
table = soup.find('table')

### ... and prepare data.

We need to transform html data in a data frame with organized informations, so we will can work better with the data.

In [5]:

df = pd.read_html(str(table))
toronto_df = df[0]
toronto_df.columns = toronto_df.iloc[0, :]
toronto_df.drop(toronto_df.index[0], inplace=True)
toronto_df.drop(toronto_df.index[toronto_df['Borough'] == 'Not assigned'], inplace=True)
toronto_df = toronto_df.rename(columns = {'Postcode':'PostalCode', 'Neighbourhood':'Neighborhood'})
toronto_df2 = toronto_df.groupby('PostalCode')['Neighborhood'].apply(', '.join)
toronto_df2 = pd.DataFrame(toronto_df2).reset_index()
toronto_df3 = toronto_df.groupby('PostalCode')['Borough'].first()
toronto_df3 = pd.DataFrame(toronto_df3).reset_index()
toronto_df3['Neighborhood'] = toronto_df2['Neighborhood']
toronto_df3.iloc[toronto_df3.index[toronto_df3['Neighborhood'] == 'Not assigned'],2] = toronto_df3.iloc[toronto_df3.index[toronto_df3['Neighborhood'] == 'Not assigned'],1]
toronto_df3


PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hill, Port Union  
2                    Guildwood, Morningside, West Hill  
3                                               Woburn  
4                                            Cedarbrae  
5                                  Scarborough Village  
6          East Birchmount Park, Ionview, Kennedy Park  
7                      Clairlea, Golden Mile, Oakridge  
8      Cliffcrest, Cliffside, Scarborough Village West  
9                          Birch Cliff, Cliffside West  
10   Dorset Park, Scarborough Town Centre, Wexford ...  
11                                   Maryvale, Wexford  
12                                           Agincourt  
13             Clarks Corners, Sullivan, Tam O'Shanter  
14   Agincourt North, L'Amoreaux East, Milliken, St...  
15                                     L'Amoreaux West  
16                                         Upper Rouge  
17                                   Hillcrest Village  
18                        Fairview, Henry Farm, Oriole  
19                                     Bayview Village  
20                            Silver Hills, York Mills  
21                             Newtonbrook, Willowdale  
22                                    Willowdale South  
23                                     York Mills West  
24                                     Willowdale West  
25                                           Parkwoods  
26                                     Don Mills North  
27                    Flemingdon Park, Don Mills South  
28     Bathurst Manor, Downsview North, Wilson Heights  
29                     Northwood Park, York University  
..  

In [6]:
# discovering how many rows and columns has the table

print(toronto_df3.shape)

(103, 3)


### 3.2. Appending Latitude/Longitude to neighborhoods data from Toronto

Here we get the latitude and longitude data for the postal codes from our data source.

In [7]:
# get the latlong data for postal codes

latlong = pd.read_csv('http://cocl.us/Geospatial_data')
latlong.shape

(103, 3)

In [8]:
latlong

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
5           M1J  43.744734 -79.239476
6           M1K  43.727929 -79.262029
7           M1L  43.711112 -79.284577
8           M1M  43.716316 -79.239476
9           M1N  43.692657 -79.264848
10          M1P  43.757410 -79.273304
11          M1R  43.750072 -79.295849
12          M1S  43.794200 -79.262029
13          M1T  43.781638 -79.304302
14          M1V  43.815252 -79.284577
15          M1W  43.799525 -79.318389
16          M1X  43.836125 -79.205636
17          M2H  43.803762 -79.363452
18          M2J  43.778517 -79.346556
19          M2K  43.786947 -79.385975
20          M2L  43.757490 -79.374714
21          M2M  43.789053 -79.408493
22          M2N  43.770120 -79.408493
23          M2P  43.752758 -79.400049
24          M2R  43.782736 -79.442259
25          M3A  43.753259 -79.329656
26          M3B  43.745906 -79.352188
27          M3C  43.725900 -79.340923
28          M3H  43.754328 -79.442259
29          M3J  43.767980 -79.487262
..          ...        ...        ...
73          M6C  43.693781 -79.428191
74          M6E  43.689026 -79.453512
75          M6G  43.669542 -79.422564
76          M6H  43.669005 -79.442259
77          M6J  43.647927 -79.419750
78          M6K  43.636847 -79.428191
79          M6L  43.713756 -79.490074
80          M6M  43.691116 -79.476013
81          M6N  43.673185 -79.487262
82          M6P  43.661608 -79.464763
83          M6R  43.648960 -79.456325
84          M6S  43.651571 -79.484450
85          M7A  43.662301 -79.389494
86          M7R  43.636966 -79.615819
87          M7Y  43.662744 -79.321558
88          M8V  43.605647 -79.501321
89          M8W  43.602414 -79.543484
90          M8X  43.653654 -79.506944
91          M8Y  43.636258 -79.498509
92          M8Z  43.628841 -79.520999
93          M9A  43.667856 -79.532242
94          M9B  43.650943 -79.554724
95          M9C  43.643515 -79.577201
96          M9L  43.756303 -79.565963
97          M9M  43.724766 -79.532242
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

And here we append the latitude and longitude data to corresponding postal codes.

In [9]:
toronto_df3['Latitude'] = latlong['Latitude']
toronto_df3['Longitude'] = latlong['Longitude']
toronto_df3

PostalCode           Borough  \
0          M1B       Scarborough   
1          M1C       Scarborough   
2          M1E       Scarborough   
3          M1G       Scarborough   
4          M1H       Scarborough   
5          M1J       Scarborough   
6          M1K       Scarborough   
7          M1L       Scarborough   
8          M1M       Scarborough   
9          M1N       Scarborough   
10         M1P       Scarborough   
11         M1R       Scarborough   
12         M1S       Scarborough   
13         M1T       Scarborough   
14         M1V       Scarborough   
15         M1W       Scarborough   
16         M1X       Scarborough   
17         M2H        North York   
18         M2J        North York   
19         M2K        North York   
20         M2L        North York   
21         M2M        North York   
22         M2N        North York   
23         M2P        North York   
24         M2R        North York   
25         M3A        North York   
26         M3B        North York   
27         M3C        North York   
28         M3H        North York   
29         M3J        North York   
..         ...               ...   
73         M6C              York   
74         M6E              York   
75         M6G  Downtown Toronto   
76         M6H      West Toronto   
77         M6J      West Toronto   
78         M6K      West Toronto   
79         M6L        North York   
80         M6M              York   
81         M6N              York   
82         M6P      West Toronto   
83         M6R      West Toronto   
84         M6S      West Toronto   
85         M7A      Queen's Park   
86         M7R       Mississauga   
87         M7Y      East Toronto   
88         M8V         Etobicoke   
89         M8W         Etobicoke   
90         M8X         Etobicoke   
91         M8Y         Etobicoke   
92         M8Z         Etobicoke   
93         M9A         Etobicoke   
94         M9B         Etobicoke   
95         M9C         Etobicoke   
96         M9L        North York   
97         M9M        North York   
98         M9N              York   
99         M9P         Etobicoke   
100        M9R         Etobicoke   
101        M9V         Etobicoke   
102        M9W         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                       Rouge, Malvern  43.806686 -79.194353  
1               Highland Creek, Rouge Hill, Port Union  43.784535 -79.160497  
2                    Guildwood, Morningside, West Hill  43.763573 -79.188711  
3                                               Woburn  43.770992 -79.216917  
4                                            Cedarbrae  43.773136 -79.239476  
5                                  Scarborough Village  43.744734 -79.239476  
6          East Birchmount Park, Ionview, Kennedy Park  43.727929 -79.262029  
7                      Clairlea, Golden Mile, Oakridge  43.711112 -79.284577  
8      Cliffcrest, Cliffside, Scarborough Village West  43.716316 -79.239476  
9                          Birch Cliff, Cliffside West  43.692657 -79.264848  
10   Dorset Park, Scarborough Town Centre, Wexford ...  43.757410 -79.273304  
11                                   Maryvale, Wexford  43.750072 -79.295849  
12                                           Agincourt  43.794200 -79.262029  
13             Clarks Corners, Sullivan, Tam O'Shanter  43.781638 -79.304302  
14   Agincourt North, L'Amoreaux East, Milliken, St...  43.815252 -79.284577  
15                                     L'Amoreaux West  43.799525 -79.318389  
16                                         Upper Rouge  43.836125 -79.205636  
17                                   Hillcrest Village  43.803762 -79.363452  
18                        Fairview, Henry Farm, Oriole  43.778517 -79.346556  
19                                     Bayview Village  43.786947 -79.385975  
20                            Silver Hills, York Mills  43.757490 -79.374714  
21                             Ne

### 3.3. Exploring the neighborhoods in Toronto

We will visualize the neighborhoods in a map of Toronto.

In [10]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

!conda install -c conda-forge geopy --yes # installing the conda forge package
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

print('Libraries imported.')

Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    geographiclib: 1.49-py_0   conda-forge
    geopy:         1.19.0-py_0 conda-forge

geographiclib- 100% |################################| Time: 0:00:00  23.91 MB/s
geopy-1.19.0-p 100% |################################| Time: 0:00:00  36.59 MB/s
Fetching package metadata .............
Solving package specifications: .

Package plan for installation in environment /opt/conda/envs/DSX-Python35:

The following NEW packages will be INSTALLED:

    altair:  2.2.2-py35_1 conda-forge
    branca:  0.3.1-py_0   conda-forge
    folium:  0.5.0-py_0   conda-forge
    vincent: 0.4.4-py_1   conda-forge

altair-2.2.2-p 100% |################################| Time: 0:00:00  48.98 MB/s
branca-0.3.1-p 100% |################################| Time: 0:00:00  37.47 MB/s
vincent-0.4.4- 100% |###################

Use geopy library to get the latitude and longitude values of Toronto.

In [11]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


Create a map of Toronto with neighborhoods superimposed on top.

In [12]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df3['Latitude'], toronto_df3['Longitude'], toronto_df3['Borough'], toronto_df3['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

### 3.4. Now we will use Foursquare to request a exploratory search of venues in the neighborhoods.

Define Foursquare Credentials and Version

In [13]:
# The code was removed by Watson Studio for sharing.

Your credentails:
CLIENT_ID: V2J0YSPY4SV5MAX5YJXRUN14CYSWQ4HHZGOTIXVUEN4YUWSO
CLIENT_SECRET:NRWNMBCXRN14QC4DQXZEVQ5TEZYTOPV4QDEDELMRXXFAYW1F


Here we explore the first neighborhood in our dataframe.

First we will get the first neighborhood's name.

In [14]:
# Get the neighborhood's name.

toronto_df3.loc[0, 'Neighborhood']

'Rouge, Malvern'

Get the neighborhood's latitude and longitude values.

In [15]:
neighborhood_latitude = toronto_df3.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = toronto_df3.loc[0, 'Longitude'] # neighborhood longitude value

neighborhood_name = toronto_df3.loc[0, 'Neighborhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rouge, Malvern are 43.806686299999996, -79.19435340000001.


Now, we will get the top 100 venues that are in Marble Hill within a radius of 500 meters.

First, let's create the GET request URL. Name your URL **url**.

In [16]:
LIMIT = 100
radius = 500

url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)

url

'https://api.foursquare.com/v2/venues/explore?&client_id=V2J0YSPY4SV5MAX5YJXRUN14CYSWQ4HHZGOTIXVUEN4YUWSO&client_secret=NRWNMBCXRN14QC4DQXZEVQ5TEZYTOPV4QDEDELMRXXFAYW1F&v=20180605&ll=43.806686299999996,-79.19435340000001&radius=500&limit=100'

Send the GET request and examine the resutls

In [17]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5ca0d28ddd57972612719c6c'},
 'response': {'groups': [{'items': [{'reasons': {'count': 0,
       'items': [{'reasonName': 'globalInteractionReason',
         'summary': 'This spot is popular',
         'type': 'general'}]},
      'referralId': 'e-0-4bb6b9446edc76b0d771311c-0',
      'venue': {'categories': [{'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/food/fastfood_',
          'suffix': '.png'},
         'id': '4bf58dd8d48988d16e941735',
         'name': 'Fast Food Restaurant',
         'pluralName': 'Fast Food Restaurants',
         'primary': True,
         'shortName': 'Fast Food'}],
       'id': '4bb6b9446edc76b0d771311c',
       'location': {'cc': 'CA',
        'city': 'Toronto',
        'country': 'Canada',
        'crossStreet': 'Morningside & Sheppard',
        'distance': 387,
        'formattedAddress': ['Toronto ON', 'Canada'],
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.80744841934756,
          'ln

In [18]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

Now we are ready to clean the json and structure it into a *pandas* dataframe.

In [19]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name            categories        lat        lng
0  Wendy's  Fast Food Restaurant  43.807448 -79.199056

And how many venues were returned by Foursquare?

In [20]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

1 venues were returned by Foursquare.


### 3.5. Explore all neighborhoods in Toronto

Here we create a function to repeat the same process to all the neighborhoods in Toronto

In [21]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Now we write the code to run the above function on each neighborhood and create a new dataframe called *toronto_venues*.

In [22]:
toronto_venues = getNearbyVenues(names=toronto_df3['Neighborhood'],
                                   latitudes=toronto_df3['Latitude'],
                                   longitudes=toronto_df3['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam O'Shanter
Agincourt North, L'Amoreaux East, Milliken, Steeles East
L'Amoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, 

Let's check the size of the resulting dataframe

In [23]:
print(toronto_venues.shape)
toronto_venues.head()

(2235, 7)


Neighborhood  Neighborhood Latitude  \
0                          Rouge, Malvern              43.806686   
1  Highland Creek, Rouge Hill, Port Union              43.784535   
2       Guildwood, Morningside, West Hill              43.763573   
3       Guildwood, Morningside, West Hill              43.763573   
4       Guildwood, Morningside, West Hill              43.763573   

   Neighborhood Longitude                            Venue  Venue Latitude  \
0              -79.194353                          Wendy's       43.807448   
1              -79.160497            Royal Canadian Legion       43.782533   
2              -79.188711  Swiss Chalet Rotisserie & Grill       43.767697   
3              -79.188711                G & G Electronics       43.765309   
4              -79.188711                       Marina Spa       43.766000   

   Venue Longitude        Venue Category  
0       -79.199056  Fast Food Restaurant  
1       -79.163085                   Bar  
2       -79.189914           Pizza Place  
3       -79.191537     Electronics Store  
4       -79.191000                   Spa

Then we check how many venues were returned for each neighborhood

In [24]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Agincourt                                                               4   
Agincourt North, L'Amoreaux East, Milliken, Ste...                      2   
Albion Gardens, Beaumond Heights, Humbergate, J...                     10   
Alderwood, Long Branch                                                 10   
Bathurst Manor, Downsview North, Wilson Heights                        18   
Bayview Village                                                         4   
Bedford Park, Lawrence Manor East                                      24   
Berczy Park                                                            55   
Birch Cliff, Cliffside West                                             4   
Bloordale Gardens, Eringate, Markland Wood, Old...                      7   
Brockton, Exhibition Place, Parkdale Village                           22   
Business Reply Mail Processing Centre 969 Eastern                      17   
CFB Toronto, Downsview East                                             3   
CN Tower, Bathurst Quay, Island airport, Harbou...                     14   
Cabbagetown, St. James Town                                            46   
Caledonia-Fairbanks                                                     6   
Canada Post Gateway Processing Centre                                  11   
Cedarbrae                                                               7   
Central Bay Street                                                     83   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               16   
Church and Wellesley                                                   86   
Clairlea, Golden Mile, Oakridge                                         9   
Clarks Corners, Sullivan, Tam O'Shanter                                 9   
Cliffcrest, Cliffside, Scarborough Village West                         3   
Cloverdale, Islington, Martin Grove, Princess G...                      2   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             37   
Davisville North                                                        8   
Deer Park, Forest Hill SE, Rathnelly, South Hil...                     14   
Del Ray, Keelesdale, Mount Dennis, Silverthorn                          5   
Design Exchange, Toronto Dominion Centre                              100   
Don Mills North                                                         5   
Dorset Park, Scarborough Town Centre, Wexford H...                      7   
Dovercourt Village, Dufferin                                           18   
Downsview Central                                                       3   
Downsview Northwest                                                     5   
Downsview West                                                          4   
Downsview, North Park, Upwood Park                                      4   
East Birchmount Park, Ionview, Kennedy Park                             6   
East Toronto                                                            4   
Emery, Humberlea                                                        2   
Fairview, Henry Farm, Oriole                                           66   
First Canadian Place, Underground city                                100   
Flemingdon Park, Don Mills South                                       21   
Forest Hill North, Forest Hill West                                     4   
Glencairn                                                               4   
Guildwood, Morningside, West Hill                                       8   
Harbord, University of Toronto                                         35   
Harbourfront East, Toronto Islands, Union Statio

Here we will find out how many unique categories can be curated from all the returned venues

In [25]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 273 uniques categories.


### 3.6. Analyze Each Neighborhood

For each neighborhood we will determin wich categories of venues are most common.

In [26]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.drop(['Neighborhood'], axis=1, inplace=True)

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  \
0                          Rouge, Malvern                  0               0   
1  Highland Creek, Rouge Hill, Port Union                  0               0   
2       Guildwood, Morningside, West Hill                  0               0   
3       Guildwood, Morningside, West Hill                  0               0   
4       Guildwood, Morningside, West Hill                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  Athletics & Sports  Auto Garage  Auto Workshop  \
0                 0                   0            0              0   
1                 0                   0            0              0   
2                 0                   0            0              0   
3                 0                   0            0              0   
4                 0                   0            0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Field  \
0          0           0           0       0     0    0               0   
1          0           0           0       0     0    1               0   
2          0           0           0       0     0    0               0   
3          0           0           0       0     0    0               0   
4          0           0           0       0     0    0               0   

   Baseball Stadium  Basketball Court  Basketball Stadium  Beach  Beer Bar  \
0                 0                 0                   0      0         0   
1                 0                 0                   0      0         0   
2                 0                 0                   0      0         0   
3                 0                 0                   0      0         0   
4                 0                 0                   0      0         0   

   Beer Store  Belgian Restaurant  Bike Shop  Bistro  Boat or Ferry  \
0           0                   0          0       0              0   
1           0                   0          0       0              0   
2           0                   0          0       0              0   
3           0                   0          0       0              0   
4           0                   0          0       0              0   

   Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  Brewery  \
0          0         0                     0               0        0   
1          0         0                     0               0        0   
2          0         0                     0               0        0   
3          0         0                     0               0        0   
4          0         0                     0               0        0   

   Bridal Shop  Bu

Examining the new dataframe size we see that have 2235 venues in the neighborhoods and 272 categories of venues.

In [27]:
toronto_onehot.shape

(2235, 273)

So we group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [28]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                            Adelaide, King, Richmond               0.01   
1                                           Agincourt               0.00   
2   Agincourt North, L'Amoreaux East, Milliken, St...               0.00   
3   Albion Gardens, Beaumond Heights, Humbergate, ...               0.00   
4                              Alderwood, Long Branch               0.00   
5     Bathurst Manor, Downsview North, Wilson Heights               0.00   
6                                     Bayview Village               0.00   
7                   Bedford Park, Lawrence Manor East               0.00   
8                                         Berczy Park               0.00   
9                         Birch Cliff, Cliffside West               0.00   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...               0.00   
11       Brockton, Exhibition Place, Parkdale Village               0.00   
12  Business Reply Mail Processing Centre 969 Eastern               0.00   
13                        CFB Toronto, Downsview East               0.00   
14  CN Tower, Bathurst Quay, Island airport, Harbo...               0.00   
15                        Cabbagetown, St. James Town               0.00   
16                                Caledonia-Fairbanks               0.00   
17              Canada Post Gateway Processing Centre               0.00   
18                                          Cedarbrae               0.00   
19                                 Central Bay Street               0.00   
20          Chinatown, Grange Park, Kensington Market               0.00   
21                                           Christie               0.00   
22                               Church and Wellesley               0.00   
23                    Clairlea, Golden Mile, Oakridge               0.00   
24            Clarks Corners, Sullivan, Tam O'Shanter               0.00   
25    Cliffcrest, Cliffside, Scarborough Village West               0.00   
26  Cloverdale, Islington, Martin Grove, Princess ...               0.00   
27                     Commerce Court, Victoria Hotel               0.00   
28                                         Davisville               0.00   
29                                   Davisville North               0.00   
30  Deer Park, Forest Hill SE, Rathnelly, South Hi...               0.00   
31     Del Ray, Keelesdale, Mount Dennis, Silverthorn               0.00   
32           Design Exchange, Toronto Dominion Centre               0.00   
33                                    Don Mills North               0.00   
34  Dorset Park, Scarborough Town Centre, Wexford ...               0.00   
35                       Dovercourt Village, Dufferin               0.00   
36                                  Downsview Central               0.00   
37                                Downsview Northwest               0.00   
38                                     Downsview West               0.00   
39                 Downsview, North Park, Upwood Park               0.00   
40        East Birchmount Park, Ionview, Kennedy Park               0.00   
41                                       East Toronto               0.00   
42                                   Emery, Humberlea               0.00   
43                       Fairview, Henry Farm, Oriole               0.00   
44             First Canadian Place, Underground city               0.00   
45                   Flemingdon Park, Don Mills South               0.00   
46                Forest Hill North, Forest Hill West               0.00   
47                                          Glencairn               0.00   
48                  Guildwood, Morningside, West Hill               0.00   
49                     Harbord, University of Toronto               0.00   
50  Harbourfront East, Toronto Islands, Union Station               0.00   
51                          Harbourfront, Regent Park               0.00   
52           

After group by neighborhood we confirm the new size of data frame.

In [29]:
toronto_grouped.shape

(98, 273)

Here we will print each neighborhood along with the top 5 most common venues.

In [30]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.06
1              Bar  0.04
2             Café  0.04
3  Thai Restaurant  0.04
4       Steakhouse  0.04


----Agincourt----
               venue  freq
0             Lounge  0.25
1     Breakfast Spot  0.25
2     Sandwich Place  0.25
3       Skating Rink  0.25
4  Accessories Store  0.00


----Agincourt North, L'Amoreaux East, Milliken, Steeles East----
                             venue  freq
0                       Playground   0.5
1                             Park   0.5
2                      Men's Store   0.0
3              Monument / Landmark   0.0
4  Molecular Gastronomy Restaurant   0.0


----Albion Gardens, Beaumond Heights, Humbergate, Jamestown, Mount Olive, Silverstone, South Steeles, Thistletown----
                  venue  freq
0         Grocery Store   0.2
1           Pizza Place   0.1
2              Pharmacy   0.1
3  Fast Food Restaurant   0.1
4          Liquor Store   0.1


----Alderwood, Long 

And put that into a *pandas* dataframe.

First, we write a function to sort the venues in descending order.

In [31]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now we will create the new dataframe and display the top 10 venues for each neighborhood.

In [32]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North, L'Amoreaux East, Milliken, St...            Playground   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Steakhouse       Thai Restaurant                   Bar   
1        Sandwich Place        Breakfast Spot          Skating Rink   
2                  Park           Yoga Studio             Drugstore   
3           Pizza Place   Fried Chicken Joint           Coffee Shop   
4          Skating Rink          Dance Studio              Pharmacy   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                  Café      Sushi Restaurant                 Hotel   
1             Drugstore        Discount Store               Dog Run   
2    Dim Sum Restaurant                 Diner        Discount Store   
3        Sandwich Place          Liquor Store  Fast Food Restaurant   
4           Coffee Shop                  Pool                   Pub   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0      Asian Restaurant   American Restaurant                       Bakery  
1      Doner Restaurant            Donut Shop  Eastern European Restaurant  
2               Dog Run      Doner Restaurant                   Donut Shop  
3            Beer Store              Pharmacy            German Restaurant  
4        Sandwich Place                   Gym             Airport Terminal

### 3.7. Clustering neighborhoods using the k-means methodology

We will use a machine learning algorithm to cluster the neighborhoods based in the categories of venues contained in each of them. This way we expect to better understand how similar are the neighborhoods and agroup them to form five clusters, so we can take a look and decide latter where a Baseball shop should be open to have the most likely public based on categories of venues more frequently in a cluster.

Run *k*-means to cluster the neighborhood into 5 clusters.

In [33]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 1, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

And now we will create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [34]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df3

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.dropna(inplace=True)
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype(int)

toronto_merged.head() # check the last columns!

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353               0  Fast Food Restaurant   Dumpling Restaurant   
1 -79.160497               3                   Bar           Yoga Studio   
2 -79.188711               0           Pizza Place                   Spa   
3 -79.216917               0           Coffee Shop     Korean Restaurant   
4 -79.239476               0      Hakka Restaurant  Caribbean Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                 Diner        Discount Store               Dog Run   
1        Discount Store               Dog Run      Doner Restaurant   
2     Electronics Store        Breakfast Spot   Rental Car Location   
3           Yoga Studio   Dumpling Restaurant        Discount Store   
4                  Bank                Bakery   Fried Chicken Joint   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Doner Restaurant            Donut Shop             Drugstore   
1            Donut Shop             Drugstore   Dumpling Restaurant   
2    Mexican Restaurant          Intersection        Medical Center   
3               Dog Run      Doner Restaurant            Donut Shop   
4       Thai Restaurant    Athletics & Sports       Coworking Space   

         9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant          Women's Store  
1  Eastern European Restaurant                  Field  
2             Doner Restaurant         Discount Store  
3                    Drugstore      Electronics Store  
4                     Creperie     Falafel Restaurant

## 4. Results

Finally, we will visualize the resulting clusters in Toronto map. This way we can see clearly where are the similar neighborhoods distributed along the city.

In [35]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

In the resulting map we can see that are a very large cluster, one medium size and other three are small clusters. In next section we will search for the most common venues categories presented in each cluster to understand the patterns that resulted in this unequal distribution.

## 5. Discussion

After show the cluster in a map, we need to examine clusters, understand what are the similarities between each group. Now, you can see each cluster and the most common venue categories contained in the neighborhoods that distinguish each cluster.

#### Cluster 1

In [37]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
0                                       Rouge, Malvern               0   
2                    Guildwood, Morningside, West Hill               0   
3                                               Woburn               0   
4                                            Cedarbrae               0   
6          East Birchmount Park, Ionview, Kennedy Park               0   
7                      Clairlea, Golden Mile, Oakridge               0   
8      Cliffcrest, Cliffside, Scarborough Village West               0   
9                          Birch Cliff, Cliffside West               0   
10   Dorset Park, Scarborough Town Centre, Wexford ...               0   
11                                   Maryvale, Wexford               0   
12                                           Agincourt               0   
13             Clarks Corners, Sullivan, Tam O'Shanter               0   
15                                     L'Amoreaux West               0   
17                                   Hillcrest Village               0   
18                        Fairview, Henry Farm, Oriole               0   
19                                     Bayview Village               0   
22                                    Willowdale South               0   
24                                     Willowdale West               0   
26                                     Don Mills North               0   
27                    Flemingdon Park, Don Mills South               0   
28     Bathurst Manor, Downsview North, Wilson Heights               0   
29                     Northwood Park, York University               0   
31                                      Downsview West               0   
32                                   Downsview Central               0   
33                                 Downsview Northwest               0   
34                                    Victoria Village               0   
35                     Woodbine Gardens, Parkview Hill               0   
36                                    Woodbine Heights               0   
37                                         The Beaches               0   
38                                             Leaside               0   
39                                    Thorncliffe Park               0   
41                        The Danforth West, Riverdale               0   
42                      The Beaches West, India Bazaar               0   
43                                     Studio District               0   
45                                    Davisville North               0   
46                                  North Toronto West               0   
47                                          Davisville               0   
49   Deer Park, Forest Hill SE, Rathnelly, South Hi...               0   
51                         Cabbagetown, St. James Town               0   
52                                Church and Wellesley               0   
53                           Harbourfront, Regent Park               0   
54                            Ryerson, Garden District               0   
55                                      St. James Town               0   
56                                         Berczy Park               0   
57                                  Central Bay Street               0   
58                            Adelaide, King, Richmond               0   
59   Harbourfront East, Toronto Islands, Union Station               0   
60            Design Exchange, Toronto Dominion Centre               0   
61                      Commerce Court, Victoria Hotel               0   
62                   Bedford Park, Lawrence Manor East               0   
63                                            Roselawn               0   
64                 Forest Hill North, Forest Hill West               0   
65                 The Annex, North Midtown, Yorkville               0   
66                      Harbord, University of

#### Cluster 2

In [38]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
5                                 Scarborough Village               1   
14  Agincourt North, L'Amoreaux East, Milliken, St...               1   
48                        Moore Park, Summerhill East               1   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
5             Playground             Drugstore    Dim Sum Restaurant   
14            Playground                  Park           Yoga Studio   
48            Playground                   Gym             Drugstore   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
5                  Diner        Discount Store               Dog Run   
14             Drugstore    Dim Sum Restaurant                 Diner   
48    Dim Sum Restaurant                 Diner        Discount Store   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5       Doner Restaurant            Donut Shop   Dumpling Restaurant   
14        Discount Store               Dog Run      Doner Restaurant   
48               Dog Run      Doner Restaurant            Donut Shop   

   10th Most Common Venue  
5      College Rec Center  
14             Donut Shop  
48    Dumpling Restaurant

#### Cluster 3

In [39]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
23                                     York Mills West               2   
25                                           Parkwoods               2   
30                         CFB Toronto, Downsview East               2   
40                                        East Toronto               2   
44                                       Lawrence Park               2   
50                                            Rosedale               2   
74                                 Caledonia-Fairbanks               2   
79                  Downsview, North Park, Upwood Park               2   
90       The Kingsway, Montgomery Road, Old Mill North               2   
100  Kingsview Village, Martin Grove Gardens, Richv...               2   

    1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
23                   Park                  Bank           Yoga Studio   
25   Fast Food Restaurant     Food & Drink Shop                  Park   
30               Bus Stop               Airport                  Park   
40                   Park           Coffee Shop     Convenience Store   
44   Gym / Fitness Center                  Park              Bus Line   
50                   Park                 Trail            Playground   
74                   Park  Fast Food Restaurant                Market   
79       Basketball Court                  Park                Bakery   
90                   Park                 River            Smoke Shop   
100           Pizza Place                  Park              Bus Line   

          4th Most Common Venue        5th Most Common Venue  \
23          Dumpling Restaurant               Discount Store   
25                        Diner               Discount Store   
30                  Yoga Studio          Dumpling Restaurant   
40                  Yoga Studio          Dumpling Restaurant   
44                  Swim School                  Yoga Studio   
50                  Yoga Studio           Dim Sum Restaurant   
74                     Pharmacy                Women's Store   
79   Construction & Landscaping  Eastern European Restaurant   
90                         Pool                  Yoga Studio   
100                   Drugstore                        Diner   

    6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
23                Dog Run      Doner Restaurant            Donut Shop   
25                Dog Run      Doner Restaurant            Donut Shop   
30         Discount Store               Dog Run      Doner Restaurant   
40         Discount Store               Dog Run      Doner Restaurant   
44              Drugstore        Discount Store               Dog Run   
50                  Diner        Discount Store               Dog Run   
74          Grocery Store      Department Store   Empanada Restaurant   
79                Dog Run      Doner Restaurant            Donut Shop   
90             Donut Shop    Dim Sum Restaurant                 Diner   
100        Discount Store               Dog Run      Doner Restaurant   

    9th Most Common Venue       10th Most Common Venue  
23              Drugstore  Eastern European Restaurant  
25              Drugstore          Dumpling Restaurant  
30             Donut Shop                    Drugstore  
40             Donut Shop                    Drugstore  
44       Doner Restaurant                   Donut Shop  
50       Doner Restaurant                   Donut Shop  
74                    Gym            Electronics Store  
79              Drugstore          Dumpling Restaurant  
90         Discount Store                      Dog Run  
100            Donut Shop          Dumpling Restaurant

#### Cluster 4

In [40]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
1  Highland Creek, Rouge Hill, Port Union               3   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
1                   Bar           Yoga Studio        Discount Store   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
1               Dog Run      Doner Restaurant            Donut Shop   

  7th Most Common Venue 8th Most Common Venue        9th Most Common Venue  \
1             Drugstore   Dumpling Restaurant  Eastern European Restaurant   

  10th Most Common Venue  
1                  Field

#### Cluster 5

In [41]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(5, toronto_merged.shape[1]))]]

Neighborhood  Cluster Labels  \
91  Humber Bay, King's Mill Park, Kingsway Park So...               4   
97                                   Emery, Humberlea               4   

   1st Most Common Venue          2nd Most Common Venue 3rd Most Common Venue  \
91        Baseball Field                           Pool           Yoga Studio   
97        Baseball Field  Paper / Office Supplies Store           Yoga Studio   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
91                 Diner        Discount Store               Dog Run   
97        Discount Store               Dog Run      Doner Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
91      Doner Restaurant            Donut Shop             Drugstore   
97            Donut Shop             Drugstore   Dumpling Restaurant   

         10th Most Common Venue  
91          Dumpling Restaurant  
97  Eastern European Restaurant

Looking at results we observe that the cluster number 1 is the largest of them, so is hard to find particular venues that determines the link to this group. In the others clusters we have less neighborhoods and is possible see more clearly what venues is more common between the group. The cluster number 2 is distinguished by playgrounds and drugstore. In the cluster number 3 is more common to a neighborhood have park, yoga studio, discount store and dog run. Having just one neighborhood in the cluster number 4, the most common venue is bars. So, that neighborhood have too some similarities with the pattern of cluster 3, but strong distinguished by the bars, justifying be alone in a different group. And least cluster, number 5, we have the baseball field as most common venue of the group and second most common is yoga studio.

By this results and based in our need to recommend where open a Baseball shop, we can properly say that the better area in the city of Toronto to open this shop is in any neighborhood of cluster number 5.

## Conclusion

In this report we can see that clustering neighborhoods of a city considering the most common venues categories is a very powerful tool to understanding how similar or dissimilar are the areas along a location of interest. For this we need to have the data about neighborhoods of a city and them latitude and longitute. With this we can call the Foursquare API to make a exploratory seach for venues categories more contained in each neighborhood. In a organized data we then use the k-means to grouping the neighborhoods based in their similarities. Based on our observations is possible see the city in a new way and so we have a easy view of where is the locations better to open a Baseball shop. As we can see the answer is cluster number 5 because there are baseball fieds as most common venues and this is a attraction for public that are potential clients for the new Baseball shop.

So we recommend the Baseball shop open in the area of cluster number 5 to have better financial expectatives, resulting in more salles and consequently revenues.

Our job permit us say that for a more accurate result is possible search for data of other Baseball shops and make linear regressions with the revenues of this shops and the categories of venues more contained in respective neighborhoods of this shops. This way would tell what venues categories contribute more on less in succes of a Baseball shop.